File: testapp/src/main/java/com/appliedrec/mrtd_reader_app/CaptureResultActivity.java
package com.appliedrec.mrtd_reader_app;

import androidx.annotation.IntegerRes;
import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.drawable.RoundedBitmapDrawable;
import androidx.core.graphics.drawable.RoundedBitmapDrawableFactory;
import androidx.core.util.Pair;
import androidx.exifinterface.media.ExifInterface;
import androidx.room.util.StringUtil;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Rect;
import android.graphics.RectF;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;

import com.appliedrec.mrtd_reader_sample.R;
import com.appliedrec.mrtd_reader_sample.databinding.ActivityCaptureResultBinding;
import com.appliedrec.mrtdreader.MRTDScanResult;
import com.appliedrec.verid.core2.Bearing;
import com.appliedrec.verid.core2.Face;
import com.appliedrec.verid.core2.IRecognizable;
import com.appliedrec.verid.core2.RecognizableFace;
import com.appliedrec.verid.core2.VerID;
import com.appliedrec.verid.core2.VerIDCoreException;
import com.appliedrec.verid.core2.VerIDImageBitmap;
import com.appliedrec.verid.core2.session.FaceCapture;
import com.appliedrec.verid.core2.session.LivenessDetectionSessionSettings;
import com.appliedrec.verid.core2.session.VerIDSessionResult;
import com.appliedrec.verid.ui2.IVerIDSession;
import com.appliedrec.verid.ui2.VerIDSession;
import com.appliedrec.verid.ui2.VerIDSessionDelegate;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.Buffer;
import java.util.Arrays;
import java.util.StringJoiner;

public class CaptureResultActivity extends AppCompatActivity implements VerIDSessionDelegate {

    public static final String EXTRA_MRTD_SCAN_RESULT = "com.appliedrec.EXTRA_MRTD_SCAN_RESULT";
    private MRTDScanResult scanResult;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityCaptureResultBinding viewBinding = ActivityCaptureResultBinding.inflate(getLayoutInflater());
        setContentView(viewBinding.getRoot());
        if (getIntent() == null) {
            finish();
            return;
        }
        scanResult = getIntent().getParcelableExtra(EXTRA_MRTD_SCAN_RESULT);
        if (scanResult == null) {
            finish();
            return;
        }
        StringJoiner stringJoiner = new StringJoiner(" ");
        for (String name : scanResult.getSecondaryIdentifiers()) {
            stringJoiner.add(name);
        }
        stringJoiner.add(scanResult.getPrimaryIdentifier());
        String name = stringJoiner.toString();
        viewBinding.nameTextView.setText(name);
        RoundedBitmapDrawable drawable = RoundedBitmapDrawableFactory.create(getResources(), scanResult.getFaceImage());
        drawable.setCornerRadius((float)drawable.getIntrinsicWidth()/8f);
        viewBinding.faceImageView.setImageDrawable(drawable);
        viewBinding.faceImageView.setOnClickListener(v -> showDocumentDetails());
        viewBinding.selfieButton.setOnClickListener(v -> compareToSelfie());
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.scan_result, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.action_show_document_details) {
            showDocumentDetails();
            return true;
        }
        return false;
    }

    private void showDocumentDetails() {
        Intent intent = new Intent(this, DocumentDetailsActivity.class);
        intent.putExtra(DocumentDetailsActivity.EXTRA_MRTD_SCAN_RESULT, scanResult);
        startActivity(intent);
    }

    private void compareToSelfie() {
        try {
            VerID verID = ((MRTDReaderApplication)getApplication()).getVerID();
            LivenessDetectionSessionSettings settings = new LivenessDetectionSessionSettings();
            VerIDSession verIDSession = new VerIDSession(verID, settings);
            verIDSession.setDelegate(this);
            verIDSession.start();
        } catch (Exception exception) {
            exception.printStackTrace();
        }
    }

    @Override
    public void onSessionFinished(IVerIDSession<?> session, VerIDSessionResult result) {
        if (result.getError().isPresent()) {
            showError(R.string.face_capture_failed);
            return;
        }
        try {
            startActivity(createFaceComparisonIntent(session.getVerID(), result));
        } catch (Exception exception) {
            showError(R.string.face_comparison_failed);
        }
    }

    private void showError(@StringRes int description) {
        new AlertDialog.Builder(this)
                .setMessage(description)
                .setPositiveButton(android.R.string.ok, null)
                .create()
                .show();
    }

    private Intent createFaceComparisonIntent(VerID verID, VerIDSessionResult result) throws Exception {
        Pair<Face,IRecognizable> facePair = detectFaceInImage(scanResult.getFaceImage());
        Bitmap documentFaceImage = cropImageToFace(scanResult.getFaceImage(), facePair.first);
        if (!result.getFirstFaceCapture(Bearing.STRAIGHT).isPresent()) {
            throw new Exception("Face not present in result");
        }
        FaceCapture faceCapture = result.getFirstFaceCapture(Bearing.STRAIGHT).get();
        float score = verID.getFaceRecognition().compareSubjectFacesToFaces(new IRecognizable[]{facePair.second}, new IRecognizable[]{faceCapture.getFace()});
        byte[] documentFaceJpeg = compressImage(documentFaceImage);
        byte[] liveFaceJpeg = compressImage(faceCapture.getFaceImage());
        Intent intent = new Intent(this, FaceComparisonActivity.class);
        intent.putExtra(FaceComparisonActivity.EXTRA_IMAGE1, documentFaceJpeg);
        intent.putExtra(FaceComparisonActivity.EXTRA_IMAGE2, liveFaceJpeg);
        intent.putExtra(FaceComparisonActivity.EXTRA_SCORE, score);
        return intent;
    }

    private byte[] compressImage(Bitmap image) throws IOException {
        try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
            image.compress(Bitmap.CompressFormat.JPEG, 75, outputStream);
            outputStream.flush();
            return outputStream.toByteArray();
        }
    }

    private Pair<Face,IRecognizable> detectFaceInImage(Bitmap image) throws Exception {
        VerID verID = ((MRTDReaderApplication)getApplication()).getVerID();
        VerIDImageBitmap verIDImage = new VerIDImageBitmap(image, ExifInterface.ORIENTATION_NORMAL);
        Face[] faces = verID.getFaceDetection().detectFacesInImage(verIDImage.createFaceDetectionImage(), 1, 0);
        if (faces.length == 0) {
            throw new Exception("Face not detected in image");
        }
        IRecognizable[] recognizables = verID.getFaceRecognition().createRecognizableFacesFromFaces(faces, verIDImage);
        return new Pair<>(faces[0], recognizables[0]);
    }

    @SuppressLint("CheckResult")
    private Bitmap cropImageToFace(Bitmap image, Face face) {
        Rect bounds = new Rect();
        face.getBounds().round(bounds);
        bounds.intersect(0, 0, image.getWidth(), image.getHeight());
        return Bitmap.createBitmap(image, bounds.left, bounds.top, bounds.width(), bounds.height());
    }
}

File: testapp/src/main/java/com/appliedrec/mrtd_reader_app/CaptureResultActivity.kt
package com.appliedrec.mrtd_reader_app

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Rect
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.annotation.StringRes
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.graphics.drawable.RoundedBitmapDrawableFactory
import androidx.lifecycle.lifecycleScope
import com.appliedrec.mrtd_reader_app.databinding.ActivityCaptureResultBinding
import com.appliedrec.mrtdreader.MRTDScanResult
import com.appliedrec.verid3.common.Bearing
import com.appliedrec.verid3.common.Face
import com.appliedrec.verid3.common.Image
import com.appliedrec.verid3.common.serialization.fromBitmap
import com.appliedrec.verid3.common.serialization.toBitmap
import com.appliedrec.verid3.facecapture.CapturedFace
import com.appliedrec.verid3.facecapture.FaceCapture
import com.appliedrec.verid3.facecapture.FaceCaptureSessionResult
import com.appliedrec.verid3.facecapture.FaceTrackingPlugin
import com.appliedrec.verid3.facecapture.LivenessDetectionPlugin
import com.appliedrec.verid3.facedetection.retinaface.FaceDetectionRetinaFace
import com.appliedrec.verid3.facerecognition.arcface.cloud.FaceRecognitionArcFace
import com.appliedrec.verid3.spoofdevicedetection.cloud.SpoofDeviceDetection
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.util.StringJoiner

class CaptureResultActivity : AppCompatActivity() {

    private var scanResult: MRTDScanResult.Success? = null
    private lateinit var faceDetection: FaceDetectionRetinaFace
    private lateinit var faceRecognition: FaceRecognitionArcFace

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        faceRecognition = FaceRecognitionArcFace(this)
        val viewBinding = ActivityCaptureResultBinding.inflate(
            layoutInflater
        )
        setContentView(viewBinding.root)
        val uri = intent?.data ?: run {
            finish()
            return
        }
        lifecycleScope.launch(Dispatchers.IO) {
            try {
                faceDetection = FaceDetectionRetinaFace.create(this@CaptureResultActivity)
                scanResult = ResultFileHelper.readScanResult(uri)
                withContext(Dispatchers.Main) {

                    val stringJoiner = StringJoiner(" ")
                    for (name in scanResult!!.secondaryIdentifiers) {
                        stringJoiner.add(name)
                    }
                    stringJoiner.add(scanResult!!.primaryIdentifier)
                    val name = stringJoiner.toString()
                    viewBinding.nameTextView.text = name
                    val drawable = RoundedBitmapDrawableFactory.create(resources, scanResult!!.faceImage)
                    drawable.cornerRadius = drawable.intrinsicWidth.toFloat() / 8f
                    viewBinding.faceImageView.setImageDrawable(drawable)
                    viewBinding.faceImageView.setOnClickListener { v: View? -> showDocumentDetails() }
                    viewBinding.selfieButton.setOnClickListener { v: View? -> compareToSelfie() }
                }
            } catch (e: java.lang.Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun onStop() {
        super.onStop()
        if (isFinishing) {
            CoroutineScope(Dispatchers.Default).launch {
                faceDetection.close()
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.scan_result, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_show_document_details) {
            showDocumentDetails()
            return true
        }
        return false
    }

    private fun showDocumentDetails() {
        val intent = Intent(this, DocumentDetailsActivity::class.java)
        intent.data = this.intent.data
        startActivity(intent)
    }

    private fun compareToSelfie() {
        try {
            val activity = this
            lifecycleScope.launch(Dispatchers.Default) {
                val faceCaptureResult = FaceCapture.captureFaces(activity) {
                    createFaceDetection = { FaceDetectionRetinaFace.create(activity) }
                    createFaceTrackingPlugins = { listOf(
                        LivenessDetectionPlugin(arrayOf(
                            SpoofDeviceDetection(activity)
                        )) as FaceTrackingPlugin<Any>
                    )}
                }
                when (faceCaptureResult) {
                    is FaceCaptureSessionResult.Success -> {
                        val faceCapture = faceCaptureResult.capturedFaces.first { it.bearing == Bearing.STRAIGHT }
                        val intent = createFaceComparisonIntent(faceCapture)
                        withContext(Dispatchers.Main) {
                            startActivity(intent)
                        }
                    }
                    is FaceCaptureSessionResult.Failure -> {
                        withContext(Dispatchers.Main) {
                            showError(R.string.face_comparison_failed)
                        }
                    }
                    else -> {}
                }
            }
        } catch (exception: Exception) {
            showError(R.string.failed_to_start_face_capture_session)
        }
    }

    private fun showError(@StringRes description: Int) {
        AlertDialog.Builder(this)
            .setMessage(description)
            .setPositiveButton(android.R.string.ok, null)
            .create()
            .show()
    }

    @Throws(Exception::class)
    private suspend fun createFaceComparisonIntent(capturedFace: CapturedFace): Intent {
        val documentFaceImage = scanResult?.faceImage?.let { Image.fromBitmap(it) } ?: throw Exception("Missing document face image")
        val documentFace = faceDetection.detectFacesInImage(documentFaceImage, 1).firstOrNull()
            ?: throw Exception("Face not detected in image")
        val documentFaceCroppedImage = cropImageToFace(documentFaceImage.toBitmap(), documentFace)
        val capturedFaceTemplate = faceRecognition.createFaceRecognitionTemplates(
            listOf(capturedFace.face), capturedFace.image
        ).first()
        val documentFaceTemplate = faceRecognition.createFaceRecognitionTemplates(
            listOf(documentFace), documentFaceImage
        ).first()
        val score = faceRecognition.compareFaceRecognitionTemplates(listOf(capturedFaceTemplate), documentFaceTemplate).first()
        val documentFaceJpeg = compressImage(documentFaceCroppedImage)
        val croppedCaptureFaceImage = cropImageToFace(capturedFace.image.toBitmap(), capturedFace.face)
        val liveFaceJpeg = compressImage(croppedCaptureFaceImage)
        val intent = Intent(this, FaceComparisonActivity::class.java)
        intent.putExtra(FaceComparisonActivity.EXTRA_IMAGE1, documentFaceJpeg)
        intent.putExtra(FaceComparisonActivity.EXTRA_IMAGE2, liveFaceJpeg)
        intent.putExtra(FaceComparisonActivity.EXTRA_SCORE, score)
        intent.putExtra(FaceComparisonActivity.EXTRA_THRESHOLD, faceRecognition.defaultThreshold)
        return intent
    }

    @Throws(IOException::class)
    private fun compressImage(image: Bitmap): ByteArray {
        return ByteArrayOutputStream().use { outputStream ->
            image.compress(Bitmap.CompressFormat.JPEG, 75, outputStream)
            outputStream.flush()
            outputStream.toByteArray()
        }
    }

    @SuppressLint("CheckResult")
    private fun cropImageToFace(image: Bitmap, face: Face): Bitmap {
        val bounds = Rect()
        face.faceAspectRatio = 4/5f
        face.bounds.round(bounds)
        bounds.intersect(0, 0, image.width, image.height)
        return Bitmap.createBitmap(image, bounds.left, bounds.top, bounds.width(), bounds.height())
    }
}